### PR TITLE
Make ocamlyacc and make_opcodes ignore \r on Unix

### PR DESCRIPTION
**Pull request by dra27**
It's not possible to compile a Cygwin version of OCaml checked out using Git-for-Windows (at least with defaults) because:

1. make_opcodes doesn't like `\r\n` in header files it reads
2. ocamlyacc doesn't like escaped newline in string literals when the newline is `\r\n`

This GPR attempts to fix both of these - the change to ocamlyacc causes it to ignore `\r` in general, and I'm not 100% certain that that's a compatible thing to do.

### CODE REVIEW

**Comment by nojb** [comment]
Why not set the `in_channel` (resp the `FILE*`) to `text mode` ?

**Comment by dra27** [comment]
That doesn't work on Unix, does it? You'd still get the `\r` in the result?

**Comment by nojb** [comment]
Yes, you are right.

File: Changes [hunk 1]
```diff
@@ -366,6 +366,9 @@ Next version (4.05.0):
 - GPR#1019: Fix fatal error in Flambda mode "[functions] does not map set of
   closures ID" (Pierre Chambart, code review by Mark Shinwell and Leo White)
 
+- GPR#1043: Allow Windows CRLF line-endings in ocamlyacc on Unix.
```
Comment (damiendoligez) [inline Changes]: I'd write "on Unix and Cygwin", otherwise it's hard to see the point.

**Review by damiendoligez — APPROVED** [review]
I think it's OK for ocamlyacc to ignore all \r.

**Comment by dra27** [comment]
Thanks - I'll amend Changes and merge.

**Comment by xavierleroy** [comment]
If you want to remove `\r` at end of line only, something like this could do (yacc/reader.c, line 105):
```
        if (c == '\n') {
          if (i >= 2 && line[i-2] == '\r') {
            line[i-2] = '\n'; i--;
          }
          line[i] = '\0'; cptr = line; return;
        }
```

**Comment by dra27** [comment]
@xavierleroy - would you prefer it to be that way, or is the (fractionally) simpler "ignore them all" OK?

**Comment by dra27** [comment]
Actually, thinking further, ignoring all `\r` could result in character offsets within a line being misreported, I think, where ignoring it only at EOL should not result in that?

**Comment by xavierleroy** [comment]
I don't mind either way.  Ignoring `\r` only before `\n` is more cricket (I think this is what the Windows C run-time library does for files opened in text mode).  On the other hand, I'm positive our `.mly` files contain no `\r` in the middle of lines, and quite sure it's also the case for other users of ocamlyacc.

**Comment by dra27** [comment]
The cricket argument is compelling :wink:

**Comment by xavierleroy** [comment]
This looks very good to me.  I support merging this code in master.  @damiendoligez do you want it in 4.05 as well?

**Comment by damiendoligez** [comment]
OK for trunk and 4.05.

**Comment by xavierleroy** [comment]
Merged and cherry-picked on 4.05.  Hope it worked.

**Pull request merged**